### DISCUSSION TITLE
Strange upgrade issue

### DISCUSSION THREAD

**Discussion by Steve0212a** [wixtoolset/issues#8851] [Questions]
I have a WIX v4.05 bundle that has an MSI and some prerequisites in it.  For this, I am only talking about the WIX MSI.  I do a dotnet publish and publish to a folder with our compiled code.  I then use HarvestDirectory in the WIX MSI WixProj to build an MSI.

My issue is that sometimes after the upgrade, it is missing files.  Clean installs have no issue - it is only on upgrades.  Further, it is not on all upgrades, but only happens sometimes.  In my last test, it was missing System.Data.SqlClient.dll.  The following lines from the log contain either the component id or the file name.

```
MSI (s) (E4:D4) [10:18:57:411]: Disallowing installation of component: {437EC653-4B3E-52F0-95A5-0277BD4F4DF6} since the same component with higher versioned keyfile exists
.
.
MSI (s) (E4:AC) [10:20:07:829]: Executing op: UnregisterSharedComponentProvider(Component={437EC653-4B3E-52F0-95A5-0277BD4F4DF6},ProductCode={1123FEF6-7DED-4FBA-B1DD-0AE13BD16E62})
MSI (s) (E4:AC) [10:20:07:829]: Executing op: ComponentUnregister(ComponentId={437EC653-4B3E-52F0-95A5-0277BD4F4DF6},,BinaryType=1,)
.
.
MSI (s) (E4:AC) [10:20:21:452]: Executing op: FileRemove(,FileName=System.Data.SqlClient.dll,,ComponentId={437EC653-4B3E-52F0-95A5-0277BD4F4DF6})
MSI (s) (E4:AC) [10:20:21:454]: Verifying accessibility of file: System.Data.SqlClient.dll
.
.
MSI (s) (E4:D4) [10:21:38:442]: Executing op: ComponentRegister(ComponentId={437EC653-4B3E-52F0-95A5-0277BD4F4DF6},KeyPath=<redacted path>\System.Data.SqlClient.dll,State=3,,Disk=1,SharedDllRefCount=0,BinaryType=1)
```

My guess is that what is happening is that it thinks it should not install the file ("Disallowing installation of component"), but then still removes the old file.  I did some googling and came across https://stackoverflow.com/a/35524424/336681 which says to move "RemoveExistingProducts" before costing.  I tried adding this to my InstallExecuteSequence:

`<RemoveExistingProducts Before="CostInitialize" />`

but got this error message.  I tried setting overridable, but then it was overridden and didn't fix my issue.

`error WIX0170: The InstallExecuteSequence table contains an action 'RemoveExistingProducts' that is declared in two different locations. Please remove one of the actions or set the Overridable='yes' attribute on one of their elements.`

That compile error led me to this discussion: https://github.com/orgs/wixtoolset/discussions/8100 which says it won't be fixed until WIX v6.

I also thought this would fix it (already had that in our WXS), but it did not and still said the component would be disallowed.

`<Property Id="REINSTALLMODE" Value="amus" />`

Sorry for the long winded explanation.  Is there any way that I can do an upgrade and tell it to overwrite files even if they are older or to move the remove before costing (assuming that would fix it)?

thanks

**Comment by Steve0212a** [comment]
I was out on vacation last week.  Is there anyone that has any ideas how I can fix this?

**Reply by Steve0212a** [reply]
Can anyone help with this please?

**Reply by wmanning** [reply]
Hard to say with the limited information you provided.
I don't like heat since it's auto generation and hard to say what's going on.

MSI does not downgrade components of version files (exe, dll, ...), at least by default.  
- I think you can override with amus or similar like you tried.

>error WIX0170: The InstallExecuteSequence table contains an action 'RemoveExistingProducts' that is declared in two different locations.
- see: https://stackoverflow.com/questions/25485061/the-installexecutesequence-table-contains-an-action-removeexistingproducts-tha
- ignore the [Overridable](https://wixtoolset.org/docs/fivefour/#wix-v5-language-changes) part in the error message, that was something added to wix 5 and not likely what you need to be concerned with

**Reply by Steve0212a** [reply]
I actually thought I provided a lot of information by getting the log entries, etc. - LOL.  In this scenario, as I mentioned, "amus" is not working to resolve it.

I looked at the stack overflow article you provided. What should I put for the schedule attribute to resolve my issue.  I don't see an option in the docs (https://wixtoolset.org/docs/schema/wxs/majorupgrade/) to do RemoveExistingProducts before CostInitialize which is what was mentioned in other articles about this issue.

**Reply by wmanning** [reply]
Sorry, I see the article I linked is not 100%.
So just to be clear:
(1) Get rid of this:  ```<RemoveExistingProducts Before="CostInitialize" />```
(2) Add below:
```XML
<MajorUpgrade Schedule="afterInstallInitialize" />
```
Can't have both I think. MajorUpgrade is syntactic sugar for that IIRC.

I don't know if it'll solve you issue but [afterInstallInitialize](https://wixtoolset.org/docs/schema/wxs/majorupgrade/) will remove the existing product (aka install) before installing the new product. So it's similar to installing on a clean slate. 

It's what I see recommended if [component rules](https://robmensching.com/blog/posts/2003/10/18/component-rules-101/) are not followed.  Not sure if that's your issue, just FYI.

**Reply by Steve0212a** [reply]
Unfortunately, that did not solve it.  same log messages (higher versioned keyfile exists).  Any other ideas?  All I can think of is to force it to install in a different folder each time.  That way, it would never think it was overwriting a newer version of the file.  Thoughts?

**Reply by wmanning** [reply]
Yeah your idea of installing into different directory is an option: https://stackoverflow.com/questions/15573151/wix-force-installation-of-a-component-with-lower-version-number/15573802

**Reply by Steve0212a** [reply]
Can you think of any other way to fix it besides that before wix 6 since this should be fixed in wix 6 per https://github.com/orgs/wixtoolset/discussions/8100?  I would like to do different folders as a last resort.

thanks for the help

**Reply by wmanning** [reply]
I think that's a different issue than what you are experiencing.

I would suggest taking a look at the version of the file in the original installer (for example System.Data.SqlClient.dll) and the version in the new installer and see if the newer one has a lower version than the older one.  If so then it seems the link above is the only choice you have.  If not then it may be some other issue.

**Reply by Steve0212a** [reply]
yes - the file in the original installer is a newer version - that is why it will not overwrite it.  I am shocked that we updated all nuget packages and ended up with an older file in our new software - gotta love it...
